Clarify/solidify extensions (#80)
substrait-io/substrait: site/docs/serialization/binary_serialization.md

# Binary Serialization

Substrait can be serialized into a [protobuf](https://developers.google.com/protocol-buffers) based binary representation. The proto schema/idl files can be found on [GitHub](https://github.com/substrait-io/substrait/tree/main/binary). Proto files are place in the `io.substrait` namespace for C++/Java and the `Substrait.Protobuf` namespace for C#. 

## Plan

The main top-level object used to communicate a Substrait plan using protobuf is a Plan message. The plan message is composed of a set of data structures that minimize repetition in the serialization along with one (or more) Relation trees. 

=== "Plan Message"

    ```proto
%%% proto.message.Plan %%%
    ```

## Extensions
Protobuf supports both [simple](/extensions/#simple-extensions) and [advanced](/extensions/#advanced-extensions) extensions. Simple extensions are declared at the plan level and advanced extensions are declared at multiple levels of messages within the plan.

### Simple Extensions

> For simple extensions, a plan references the URIs associated with the simple extensions to provide additional plan capabilities. These URIs will list additional relevant information for the plan. 
> 
> Since references to simple extensions are used extensively in most plans, the serialized representation is normalized within the plan message type minimize space and serialization costs. For each simple extension YAML referenced, the plan will defined an ExtensionURI object and in-plan surrogate key that can be used in each simple extension extension definition. 

Reviewer: ```suggestion
For simple extensions, a plan references the URIs associated with the simple extensions to provide additional plan capabilities. These URIs will list additional relevant information for the plan.  For each simple extension YAML referenced, the plan will define an ExtensionURI object and in-plan surrogate key which is used to define specific extensions. 
```

Author: Did you not like the second section? Everyone who has encountered this pattern has seem so confused by the pattern than I added this section. Or do you think it was just covered well enough elsewhere?

Reviewer: I think the section is ok.  I proposed removing the sentence "Since references to simple extensions are used extensively in most plans, the serialized representation is normalized within the plan message type minimize space and serialization costs."

> Since references to simple extensions are used extensively in most plans

Conjecture?  Also, might worry/confuse people that aren't using simple extensions when they get started.

> the serialized representation is normalized within the plan message type minimize space and serialization costs.

Typo, and also the rationale could be confusing.  I think once people understand the "what" the "why" will be pretty obvious.  I'd be fine with this part coming back as I could be convinced the "why" helps in understanding the "what".

Author: > Conjecture? 

Since all functions are considered simple extensions, it seems like everyone will be using them (unless you make a plan with zero functions). I'm not sure what is the clear way to say that. Thoughts?

Reviewer: Ah, I see.  Perhaps we need to pick some terms.

"Surrogate key" is jargony although maybe apropos for databases and "pointer" feels a bit out of place.  Elsewhere in the spec/serialization domain this concept would be called referneces ([JSON](https://json-spec.readthedocs.io/reference.html)), key/keyref ([XML](https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/ms256478(v=vs.100))), anchor/reference ([YAML](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/)).  So I think there is some agreement on "reference" at least and that is the term I would expect.

That being said, you already use "reference" for the name of the first section but maybe "declaration" would be a better term.  So something like...

```
Extensions within a plan are split into two parts, a declaration and a number of references.  The declaration is encapsulated in the top level `extensions` section of the plan.  An extension declaration includes a URI which should be a unique identifier for the extension and should ideally point to a YAML document describing the extension.  An extension declaration also includes a key, this is a plan-specific value that the producer creates as an anchor/key to be referenced elsewhere.

An extension declaration is then referenced throughout the body of the plan itself.  The exact structure of these references will depend on the extension point being used but they will always include the extension's key.  For example, all scalar functions contain references to an extension which defines the semantics of the function.
```

Author: I did a pass based on your ideas/text. LMK your thoughts.

> === "Advanced Extension Protobuf"

Reviewer: ```suggestion
=== "Simple Extension Protobuf"
```

    ```proto
%%% proto.message.SimpleExtensionURI %%%
    ```

Once the YAML file is defined with a surrogate key, the surrogate key will be referenced by zero or more `SimpleExtensionDefinition`s. For each simple extension definition, a surrogate key is defined for that specific extension entity. This surrogate key is then pointed to within lower-level primitives (functions, etc) to reference that specific extension. Message properties are named `*_surrogate_key` where the surrogate key is defined and `*_pointer` when referencing the surrogate key. For example `function_surrogate_key` and `function_pointer`.

> === "Advanced Extension Protobuf"

Reviewer: ```suggestion
=== "Simple Extension Definition Protobuf"
```

    ```proto
%%% proto.message.SimpleExtensionDefinition %%%
    ```

!!! note
  Surrogate keys only have meaning within a single plan for the purposes of plan size reduction. They are not some form of global identifier. Different plans may use different surrogate keys for the same specific functions, types, type variations, etc.

!!! note
  It is valid for a plan to include `SimpleExtensionURI`s and/or `SimpleExtensionDefinition`s that are not referenced directly. It is an optimization to keep this list to the minimum set required.

> ### Advanced Extensions

Reviewer: Is there more going in this section eventually?

Author: Yes, definitely. We should add examples and whatnot. In part, this patch introduces the distinction between simple and advanced extensions which felt like an improvement in how we can explain these to people. I thought it was worthwhile getting it up with this limited advanced section to at least give people a framework for how to think about things.

Substrait protobuf exposes a special object in multiple places in the representation to expose extension capabilities. Extensions are done via this object. Extensions are separated into main concepts: 

| Advanced Extension Type | Description                                                  |
| ----------------------- | ------------------------------------------------------------ |

> | Optimization            | A change to the plan that may help some consumers work more efficiently with the plan. These properties should be propagated through plan pipelines where possible but do not impact the meaning of the plan. A consumer can safely ignore some/all of these properties. |

Reviewer: ```suggestion
| Optimization            | A change to the plan that may help some consumers work more efficiently with the plan. These properties should be propagated through plan pipelines where possible but do not impact the meaning of the plan. A consumer can safely ignore these properties. |
```

I'm not sure why you said *some*/all.  Are there examples of optimization extensions that can't be ignored?

* High Quality well supported and idiomatic bindings/compilers for key languages (Python, Javascript, C++, Go, Rust, Java)
* Compact serial representation

> The primary formats that exist that roughly qualify under these requirements include: Protobuf, Thrift, Flatbuf, Avro, Cap'N'Proto. The choise was made to Protobuf, primarily due to its clean typing system and large number of high quality language bindings. 

Reviewer: ```suggestion
The primary formats that exist that roughly qualify under these requirements include: Protobuf, Thrift, Flatbuf, Avro, Cap'N'Proto. Protobuf was chosen due to its clean typing system and large number of high quality language bindings. 
```

> The binary serialization IDLs can be found on [GitHub](https://github.com/substrait-io/substrait/tree/main/binary) as well as sampled throughout the doumentation.

Reviewer: ```suggestion
The binary serialization IDLs can be found on [GitHub](https://github.com/substrait-io/substrait/tree/main/binary) and are sampled throughout the documentation.
```
